[discussion] pool ssl/tls connections - alternative 2

Tieske opened the pull request:
Here's the second option for safe keep-alive across ssl + proxy connections.

It implements a new connect method that does it all at once (connect, ssl-handshake, proxy)

This is (imo the better) alternative to https://github.com/ledgetech/lua-resty-http/pull/202 .

There are no tests since this is very hard to test without a proxy server. Yet tests are available in the originating code here: https://github.com/Kong/kong-plugin-aws-lambda/compare/better-connect?expand=1 , which includes a forward-proxy and full integration tests.

Input is appreciated  @hamishforbes @pintsized @bungle .

replaces #201 
fixes #161 
replaces #205

pintsized commented on lib/resty/http.lua:
```diff
@@ -174,8 +174,23 @@ function _M.ssl_handshake(self, ...)
 end
 
+do
+    local new_connect = require "resty.http_connect"
+    -- Old function signatures to support:
+    -- ok, err = httpc:connect(host, port, options_table?)
+    -- ok, err = httpc:connect("unix:/path/to/unix.sock", options_table?)
+    function _M.connect(self, options, ...)
+        if type(options) == "table" then
+            -- new all-in-one interface
+            return new_connect(self, options)
+        else
+            -- backward compatible
+            return self:old_connect(options, ...)
+        end
+    end
+end
 
-function _M.connect(self, ...)
+function _M.old_connect(self, ...)
     local sock = self.sock
     if not sock then
         return nil, "not initialized"
```
I prefer the way this is explained in the docs to the way to code reads - i.e. "all-in-one" vs "tcp-only". Can we avoid naming the original function `old_connect`? To me I think it makes sense to have `connect` dispatch to either `tcp_only_connect` or `all_in_one_connect`, based on the arg types.

Tieske commented:
@pintsized all updates done, please have another look

Tieske commented:
> I would really like to have tests around this stuff, as well as the existing forwarding proxy connection code. This was contributed code and not a use case of mine, so I've always been a bit nervous about it. I've just got travis builds working. I'm wondering if we can do something with Docker Compose to test against a forwarding proxy?

Here's what we are using (docker-compose based):
 - squid: https://github.com/Kong/kong-pongo/blob/3ccaf860e101b9967c0aafa7a7d5d9b9eea5a148/assets/docker-compose.yml#L55-L68
 - squid config: https://github.com/Kong/kong-pongo/blob/3ccaf860e101b9967c0aafa7a7d5d9b9eea5a148/assets/squid.conf
 - squid auth: https://github.com/Kong/kong-pongo/blob/3ccaf860e101b9967c0aafa7a7d5d9b9eea5a148/assets/squid.passwords
 - some docs: https://github.com/Kong/kong-pongo/blob/3ccaf860e101b9967c0aafa7a7d5d9b9eea5a148/README.md#test-dependencies
 - and the tests: https://github.com/Kong/kong-plugin-aws-lambda/blob/d4e3e6d774292c805ae9d46de5610ab78c3dd9c6/spec/plugins/aws-lambda/50-http-proxy_spec.lua

Tieske commented:
@pintsized any updates?

PS. CI seems to be in a weird state. I can't click any of the links to check what's wrong?

Tieske commented:
ping

pintsized commented:
Hey, sorry (again) for being so slow. Things are pretty hectic for me at the moment.

It's passing local tests fine, I think it's just because your PR was submitted before I got Travis working - Github isn't getting the notification or something. Of course the important tests are the ones you're hosting (thanks for sending the details over).

My only concern with all of this is the amount of code duplication. Could / should `request_uri` (i.e. the simple path) not use your connector module instead? So the user would call `set_proxy_options` as they currently do if using `request_uri`, and it would just pass this to your interface? Ditto for `connect_proxy` actually, so we keep what we have to from the old API, but utilise your module?

I appreciate it's a bit of a mess at this stage, and that your contribution here is important for correctness. If we can push the old interface towards using this new code somehow, things will be much less likely to rot.

Having said all of this, I really don't have any time to work on it at the moment, and so if we need to merge this and come back to it, I can live with that. I'll hopefully have some proper time in a month or so. At which point I'd be keen to get something like your integration tests setup for this.

Thanks for your input, and sorry for not being more responsive!

Tieske commented:
@pintsized I updated the PR. I squashed the original PR into a single commit and added/changed step-by-step, to make the short hand method use the aio connect method (and fix some bugs that popped up).

I made an effort to make commits atomic so for review it is probably easiest to review individual commits.

Can you please have another look?

pintsized commented:
Awesome man, thanks for the effort here. I'll go through it properly by the end of the week.

Tieske commented:
there was another flaky test, updated the commit and pushed again, CI is green now.

pintsized commented on lib/resty/http_connect.lua:
```diff
@@ -0,0 +1,234 @@
+                   .. ":" .. (ssl_server_name or "")
+                   .. ":" .. tostring(ssl_verify)
+                   .. ":" .. (proxy_uri or "")
+                   .. ":" .. (request_scheme == "https" and proxy_authorization or "")
```
It's probably worth creating a table and using `table.concat` here, to reduce the number of allocations.

Tieske replied:
this is a single concat operation. If you do it in a loop, then it would be multiple.

pintsized commented on README.md:
```diff
@@ -151,14 +151,53 @@ Creates the http object. In case of failures, returns `nil` and a string describ
+`syntax: ok, err = httpc:connect(aio_options_table)`
+
 `syntax: ok, err = httpc:connect(host, port, options_table?)`
 
```
Do you think perhaps we should deprecate the old interface? Seems like we could use your new interface, without ssl or proxy options, and get the same behaviour without much cost?

Tieske replied:
yes.

These days everything is ssl mostly. Doing the steps manual is causing a lot of code duplication, and is very tedious because there are so many edge cases one needs to think of. A lot of parameters to each of the `connect`, `handshake`, and `proxy` calls share a lifetime over all the three methods.

That is exactly what caused me to rewrite the whole thing, such that it would be less error prone.

pintsized reviewed (changes requested):
OK, looking good :+1: 

Aside from the minor comments that I added, my only real concerns are the test coverage and the increased API surface.

For the test coverage, the major difference seems to be that `connect_proxy` no longer has any coverage at all. Am I right in thinking this method is really only there for backwards compatibility with code using the generic interface? I personally have almost no experience with forward proxy semantics, and all of the use cases here have been contributed by others. I'm more than a little nervous about breaking things I don't understand ;)

But this leads me to wonder if we should be marking anything as deprecated whilst we're at it? I guess I'm thinking; if someone new comes to this module, are we making it clear enough what they should reach for first, since they now have even more options? Seems like we could deprecate the old `connect` signatures, and also the `connect_proxy` method?

I guess that includes `ssh_handshake` too. Originally I just tried to keep the interface similar to the other cosocket modules, but the reality is that HTTP semantics are just more complicated than that, and my sense is we should encourage everyone to use your "all in one" method, right? Can you think of use cases that aren't covered by that?

I think we're also missing test coverage for a non `https` proxy, but that appears to be true on the master branch too.

Again, thanks for all the effort here.

Tieske commented on README.md:
```diff
@@ -151,14 +151,53 @@ Creates the http object. In case of failures, returns `nil` and a string describ
+
+- TCP connect
+- SSL handshake
+- HTTP-proxy (options to be set using [set_proxy_options](#set_proxy_options))
```
I changed it a couple of times over commits, final versions reads:
```
proxy_opts,             -- proxy opts, defaults to global proxy options
```

Tieske commented on lib/resty/http.lua:
```diff
@@ -654,7 +680,7 @@ function _M.send_request(self, params)
 
     -- Format and send request
-    local req = _format_request(params)
+    local req = _format_request(self, params)
```
No, `_format_request` is a local function that doesn't live on the module table `_M`. Unless there is a reason for exporting it on the module table, it should stay as it is imo.

Tieske commented on lib/resty/http_connect.lua:
```diff
@@ -0,0 +1,234 @@
+            local no_proxy_set = {}
+            -- wget allows domains in no_proxy list to be prefixed by "."
+            -- e.g. no_proxy=.mit.edu
+            for host_suffix in ngx_re_gmatch(proxy.no_proxy, "\\.?([^,]+)") do
```
no reason, fixed

Tieske commented:
> OK, looking good 👍
> 
> Aside from the minor comments that I added, my only real concerns are the test coverage and the increased API surface.
> 
> For the test coverage, the major difference seems to be that `connect_proxy` no longer has any coverage at all. Am I right in thinking this method is really only there for backwards compatibility with code using the generic interface?

Yes

> I personally have almost no experience with forward proxy semantics, and all of the use cases here have been contributed by others. I'm more than a little nervous about breaking things I don't understand ;)
> 
> But this leads me to wonder if we should be marking anything as deprecated whilst we're at it? I guess I'm thinking; if someone new comes to this module, are we making it clear enough what they should reach for first, since they now have even more options? Seems like we could deprecate the old `connect` signatures, and also the `connect_proxy` method?
> 
> I guess that includes `ssh_handshake` too.

Yes probably all three should be deprecated, and examples updated to use the aio method

> Originally I just tried to keep the interface similar to the other cosocket modules, but the reality is that HTTP semantics are just more complicated than that, and my sense is we should encourage everyone to use your "all in one" method, right? Can you think of use cases that aren't covered by that?

Nope

> 
> I think we're also missing test coverage for a non `https` proxy, but that appears to be true on the master branch too.
> 
> Again, thanks for all the effort here.

👍

pintsized commented:
Ok, I'll merge this and rework the docs, deprecating the old connection syntax.

Tieske commented:
Nice! 🎉

Tieske commented:
@pintsized fyi: I just created an [updated version of the Kong Lambda plugin](https://github.com/Kong/kong-plugin-aws-lambda/pull/50) which has integration tests for keepalive and proxy connections. The tests in the branch run against the master branch of the http client. [Everything is green](https://travis-ci.com/github/Kong/kong-plugin-aws-lambda/jobs/490860765#L797-L822).

Just extra confirmation all seems well, since proxy tests are lacking in the lib.

pintsized commented:
Nice, good to know! I'll cut a new release soon...
